client/js/widgets/scoreboard.js

```diff
@@ -0,0 +1,297 @@
+          pScores[i] = score.map( s => isNaN(parseFloat(s)) ? 0 : parseFloat(s));
+          total = this.getTotal(pScores[i]);
+        } else {
+          pScores[i] = [];
+          total = parseFloat(score);
+          if(isNaN(total))
+            total = 0;
+        }
+        pScores[i] = pScores[i].concat(Array(numRounds).fill('')).slice(0,numRounds);
+        if(showTotals)
+          pScores[i].push(total);
+        pScores[i].unshift(seats[i].get('id')); // Temporarily use the seat id here
+      }
+
+      // Sort player scores as requested
+      if(sortField == 'total' && !showTotals) // Use default sort if no totals
+        sortField = 'index';
+      if(sortField == 'total')
+        pScores.sort((a,b) => a[a.length-1] - b[b.length-1])
+      else
+        pScores.sort((a,b) => {
+          const pa = seats.filter(x=> x.get('id') == a[0])[0].get(sortField);
+          const pb = seats.filter(x=> x.get('id') == b[0])[0].get(sortField);
+          return pa < pb ? -1 : pa > pb ? 1 : 0; // These need not be numeric
+        });
+      if(!this.get('sortAscending'))
+        pScores = pScores.reverse();
+      
+      for(let i=0; i<pScores.length; i++)
+        pScores[i][0] = widgets.get(pScores[i][0]).get('player') || '-';
+      
+    } else if(typeof seats == 'object') { // Display team scores
+      let i = 0;
+      for (const team in seats) {
+        // Get array of (arrays of) seat scores.
+        const seatScores = seats[team].map(w =>  asArray(w.get(scoreProperty)));
+
+        // Make all score arrays for this the same length, then add them element-by-element
+        const n = seatScores.reduce((max, xs) => Math.max(max, xs.length), 0);
+        pScores[i] = Array(n).fill(0).map((_,i) => this.getTotal(seatScores.map(xs => xs[i] || 0)));
+        pScores[i] = pScores[i].concat(Array(numRounds).fill('')).slice(0,numRounds);
+
+        // Add totals and team name
+        if(showTotals)
+          pScores[i].push(this.getTotal(pScores[i]));
+        pScores[i].unshift(team || '-');
+        i++
+      }
+    } else { // Should never happen.
+      console.log('Internal error: invalid seats in tableCreate');
+      return
+    }
+
+    // Finally, build the table
+    if(!this.tableDOM) {
+      this.tableDOM = document.createElement('table');
+      this.domElement.appendChild(this.tableDOM);
+    } else {
+      this.tableDOM.innerHTML = '';
+    }
+
+    // Do not use player colors if team scores are being shown.
+    let showPlayerColors = this.get('showPlayerColors') && Array.isArray(seats);
+
+    let currentRound = parseInt(this.get('currentRound'));
+    if (isNaN(currentRound) || currentRound < 1 || currentRound > numRounds)
+      currentRound = null;
+
+    if(this.get('playersInColumns')) { // Scores are in columns
+      // Compute total number of rows and columns in table
+      const numCols = pScores.length + 1;
+      const numRows = numRounds + 1 + (showTotals ? 1 : 0);
```

Comment:
ok so this is how **I** want **my** table.

![image](https://user-images.githubusercontent.com/73476656/207222492-6f25e7f3-a0d8-403d-813a-92c6234fd97a.png)

@96LawDawg made some code in https://test.virtualtabletop.io/PR-1381/Parade that made it possible.
I changed it a bit so I could do the calculation in calc( )

so my only request is to either push those two values (numCols and numRows) as CSS variables or as properties so this sort table can be done only using CSS.

This is the CSS I'm using:

```
"css": {
    ".scoreboard td": {
      "width": "calc(100%/${PROPERTY seatCount})",
      "white-space": "normal",
      "text-align": "center"
    },
    ".scoreboard td:nth-child(1)": {
      "width": "50px"
    }
  }
```

Reply:
Yeah, I agree that we should add this. But I think it should be a toggle-able property. Either for rows+columns or properties for each.

If you have problems implementing this, I can help.

Reply:
Does or css format accept ${PROPERTY dimension}.rows or ${PROPERTY dimension}.0?

But I don't see a problem in having two properties. Specially if it is position locked in the JSON editor

Reply:
The subscripted properties are not accepted.

Providing access to the size of the scoreboard seems like a useful thing. These could certainly be exported as read-only properties (or even as normal properties that are set by the scoreboard widget code). Is that what you meant @ArnoldSmith86 ?

Reply:
What I mean is that the bottom scoreboard in https://test.virtualtabletop.io/PR-1381/Parade should be the default styling/behavior of the scoreboard IMHO. Because it's easier for the game dev to plan the size of the scoreboard and for players stuff doesn't jump around as much when adding players/rounds.

It should also add `word-break: break-all` on the players so that long names can be broken into several lines.

I still think that the current default is useful for games with many players or many rounds so that should be made available with something like `autosizeColumns: false` or whatever.

The size of the scoreboard is already accessible inside the `css` property just like any other property. But I would make `autosizeColumns` or whatever it's called toggle a class on the scoreboard widget which triggers different styling (you probably do that for other properties already).

Reply:
The size is available, but not the number of columns, which is what you need to compute column width.

So to recap what I think I just heard: the default styling is to provide CSS like above (perhaps exactly that with the number of columns substituted); if the widths are not what is desired then the designer can change that CSS. That can be toggled using a widget property.

If that's correct, then the only piece of the implementation I'm not clear how to do is how to get the number of columns (players) into the CSS, which will be a class in the CSS file.

Reply:
If you don't use the `style` property of your table already, you could just do `this.tableDOM.style.cssText = '--columns:'+ numCols` in your `tableCreate` function.

Reply:
@RaphaelAlvez I don't understand how the code above works. You set the player column widths to 100%/seatCount, which should by itself fill up the width. So why don't you get an overflow and a horizontal scroll bar?

Reply:
Just added column autosizing. This simply makes *all* the columns (including the left-hand column) the same width.

There is CSS in there to allow control of the left-hand column separately, but it doesn't work. I'd be grateful if someone (@ArnoldSmith86, @flackr ?) could figure out why, and also figure out why `calc((100% - var(--firstColWidth)) / (var(--columns)-1))` can't be used to size the remaining columns.

Reply:
Width, column-width, max-width, min-width, and number of columns (whatever that css is), all have strange and complicated relationships in tables.

But the css you are looking for in the first column is 

```
".scoreboard td:first-child": {
      "min-width":"50px"
    },
```

That makes it do what you want.  Don't ask why.  Ask how.

Reply:
"50px" or "var(--firstColWidth)"?

Reply:
Also.  The even distribution of columns doesn't seem perfect after this.  Col 1 set at 50px looks great in this example, but you can see the difference in Col 2 and 3 when set to 70px.  Maybe rems is a better solution?  I don't think percentages work in tables.

![50px](https://user-images.githubusercontent.com/76912527/207428914-25572667-4996-4ba6-ae79-0138fa04651d.png)

![70px](https://user-images.githubusercontent.com/76912527/207428923-8ade48ba-49c9-46d2-8c5f-0a65ef6877da.png)

Reply:
> Also.  The even distribution of columns doesn't seem perfect after this.  Col 1 set at 50px looks great in this example, but you can see the difference in Col 2 and 3 when set to 70px.  Maybe rems is a better solution?  I don't think percentages work in tables.
> 
> ![50px](https://user-images.githubusercontent.com/76912527/207428914-25572667-4996-4ba6-ae79-0138fa04651d.png)
> 
> ![70px](https://user-images.githubusercontent.com/76912527/207428923-8ade48ba-49c9-46d2-8c5f-0a65ef6877da.png)
> 

We wasted a long time trying to make it work

Reply:
I figured out the problem here.  The data for Player 2 in the scoreboard included a -3.5 value (and a -2).  Compared to the other data values of positive single digits, it is wider.  The css won't let you go below the minimum it takes to display what is there.  When I remove the negative numbers, it works.  Intuitively, it seems like css should take that all into account and spread things out.  But that would mean something wouldn't fit.  So it prioritizes making everything fit, even if the distribution isn't even.  That makes sense.  It also means nothing is broken.